proposals/nnnn-tuple-of-value-pack-expansion.md

```diff
@@ -0,0 +1,88 @@
+# Tuple of value pack expansion
+
+* Proposal: [SE-NNNN](nnnn-tuple-of-value-pack-expansion.md)
+* Authors: [Sophia Poirier](https://github.com/sophiapoirier), [Holly Borla](https://github.com/hborla)
+* Review Manager: TBD
+* Status: **Awaiting implementation**
+* Upcoming Feature Flag: `VariadicGenerics`
+* Previous Proposal: [SE-0393](0393-parameter-packs.md)
+
+## Introduction
+
+Building upon the **Value and Type Parameter Packs** proposal [SE-0393](https://forums.swift.org/t/se-0393-value-and-type-parameter-packs/63859), this proposal enables referencing a tuple value that contains a value pack inside a pack repetition pattern.
+
+## Motivation
+
+When a tuple value contains a value pack, there is no way to reference those pack elements or pass them as a value pack function argument.
+
+Additionally, type parameter packs are only permitted within a function parameter list, tuple element, or generic argument list. This precludes declaring a type parameter pack as a function return type, type alias, or as a local variable type to permit storage. The available solution to these restrictions is to contain the value pack in a tuple, which makes it important to provide full functional parity between value packs and tuple values containing them. This proposal fills that functionality gap by providing a method to reference individual value pack elements contained within a tuple.
+
+## Proposed solution
+
+This proposal extends the functionality of pack repetition patterns to values of _abstract tuple type_, which enables an implicit conversion of an _abstract tuple value_ to its contained value pack. An _abstract tuple type_ is a tuple that has an unknown length and elements of unknown types. Its elements are that of a single type parameter pack and no additional elements, and no label. In other words, the elements of the type parameter pack are the elements of the tuple. An _abstract tuple value_ is a value of _abstract tuple type_. This proposal provides methods to individually access the dynamic pack elements of an abstract tuple value inside of a repetition pattern.
+
+### Distinction between tuple values and value packs
+
+The following example demonstrates a pack repetition pattern on a value pack and an abstract tuple value separately first, then together but with the repetition pattern operating only on the value pack, and finally with the repetition pattern operating on both the value pack and the tuple value's contained value pack together interleaved.
+
+```swift
+func example<each U>(packElements value: repeat each U, tuple: (repeat each U)) {
+  print((repeat each value))
+  print((repeat each tuple))
```

Comment:
Question: how many of these nested parens are mandatory, and which ones are written here only for clarity? It's particularly eye-catching with the line after that ends with `)))`. Is there an addition to the formal grammar of the language that can be specified in this proposal, whether here or in the Detailed Design?

Reply:
Ahhhh...good point about `print`'s current declaration. Would it be helpful, simply for clarity/explanation purposes, if you declared a variadic `func print` with a code comment that this doesn't exist (yet) in the standard library, then illustrate calling that function with values? That way, the use site illustrates this specific proposal's feature without simultaneously trying to working around a library deficiency.

Reply:
I liked this suggestion and I started making this change, and then I wasn't really happy about the fact that the current example in comments shows you the output that you get from the example code, except by making this change, there is no literal output anymore because it uses a library function that does not exist, and I think that imagining what the output would be is less informative. My inclination now is to maybe just add a comment about why the extra parenthesis are there and stick with an example that actually runs with verifiably known output. Thoughts/opinions?

Reply:
Good point: whatever you feel is most clear!